### ISSUE TITLE
[FEATURE]: parallel mapping function wrapper like jax.pmap

### ISSUE THREAD

**Issue by mayfieldmobster** [hpcaitech/ColossalAI#3651]
### Describe the feature

This feature would work the same as torch.vmap but would run across multiple processes. 

Here is a basic rundown of how it would work:
1.  The input args to the wrapped function would be split into n chunks across a given axis, where n is the number of processes in the process group.
2.  Each process would get its own chunk which it would pass to a vmaped version of the wrapped functions
3.  Finally the output of the vmap would be either gathered to a destination rank or all gathered across all ranks in a given process group

Here are some examples of it in use, cmap(colossal map) is used to wrap the function:

```python
x = torch.randn(64, 128)
y = torch.randn(64, 128)
batch_dot = cmap(torch.dot)
batch_dot(x,y)
#---------------------------------------------

class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = nn.Linear(128, 128)
        self.relu = nn.ReLU()

    def forward(self, x):
        x = self.linear(x)
        x = self.relu(x)
        return x

model = Model()
batched_model = cmap(model)
input = torch.randn(64, 128)
batched_model(input)
#---------------------------------------------

@cmap(in_dims=(0,None), dst=0)
def batch_dot(x, y):
    return torch.dot(x, y)
a = torch.randn(64, 128)
b = torch.randn(128)
out = batch_dot(a, b)  # out = None on all process except process 0
```
Here you can see cmap being used to implement a basic version of 1D tensor parallel:

```python
tensor_parallel_1D = cmap(torch.mm, in_dims=(0, None))
w = torch.randn(64,128)
x = torch.randn(128,1) 
out = tensor_parallel_1D(w.unsqueeze(1), x).squeeze(1)
out.shape  # (64, 1)
```

cmap is a user-friendly way to implement single-program multiple-data (SPMD) programs.
